Simple overhead benchmarking facilities

tcojean opened the pull request:
This is done by modifying the `solver` benchmarks with:
+ A new `overhead` boolean parameter to ask whether or not an overhead benchmark should be done. This will use a simple matrix with a single `1.0` `value,` `0.0` as `x`, and `nan` as `b` (same setup as the `ginkgo_overhead` example). 
+ A new `Overhead` LinOP which does a normal solver setup in terms of Factory parameters and generation, except that it has an `apply_impl` method which calls four empty kernels. We have multiple runtime polymorphism calls which come into play here. I am most likely forgetting many of them: 
    + generate,
    + apply,
    + system_matrix and preconditioner's apply,
    + four kernel calls (`run()` functions), ...
+ This `Overhead` LinOp can be also be used as a preconditioner.

This allows the following:
+ Benchmark any solver/preconditioner with the dummy data.
+ Benchmark the overhead LinOp (optionally with preconditioner) with the dummy data.

We get the following results currently, benchmarked on BW Unicluster with 10.000 iterations:
+ CG with dummy data,
   + generate: `0.829 us`,
   + apply (1000 iterations): `1277.557 us`.
   + apply (2000 iterations) - apply (1000 iterations): `1200.118 us`
+ bicgstab with dummy data,
   + generate: `0.814 us`,
   + apply (1000 iterations): `1259.751 us`.
   + apply (2000 iterations) - apply (1000 iterations): `1239.370 us`
+ fcg with dummy data,
   + generate: `0.837 us`,
   + apply (1000 iterations): `1447.981 us`.
   + apply (2000 iterations) - apply (1000 iterations): `1410.630 us`
+ cgs with dummy data,
   + generate: `0.849 us`,
   + apply (1000 iterations): `1001.975 us`.
   + apply (2000 iterations) - apply (1000 iterations): `881.990 us`
+ gmres with dummy data,
   + generate: `0.836 us`,
   + apply (1000 iterations): `1510.327 us`.
   + apply (2000 iterations) - apply (1000 iterations): `1485.361 us`
+ Overhead "solver" with dummy data,
   + generate: `0.788 us`,
   + apply: `0.912 us`.
+ Overhead "solver" with overhead "preconditioner" with dummy data:
   + generate: `1.479 us`,
   + apply: `2.610 us`.

hartwiganzt commented:
I like this! This might also be useful for debugging / performance engineering. Would it be possible to get the information on how often a runtime decision was taken? Obviously, the overhead of counting this would be much higher, so we may need two runs?

hartwiganzt commented on benchmark/solver/solver.cpp:
```diff
@@ -58,21 +60,25 @@ DEFINE_uint32(max_iters, 1000,
+              "A comma-separated list of preconditioners to use. "
+              "Supported values are: none, jacobi, adaptive-jacobi, parilu, "
+              "ilu, overhead");
 
```
Maybe add an example of how to measure the overhead.

tcojean commented:
I added an iteration logger to the solver benchmark. I put this as part of the warmup since we should pretty much always have warmup and that doing this simple logging of iteration counts is not a "detailed" run. We can maybe put this logger in the main benchmarking portion, but that would add a few extra virtual calls so I decided to not proceed with this.

This revealed that indeed, the solver results were accounting for 1000 iterations of a CG solver, which changes the picture significantly (1.2 ms overhead over 1000 iterations isn't too bad). I added numbers for all solvers, and also used the strategy of doing 2000 iterations - 1000 iterations, which shows that the allocations while numerous are indeed pretty much nothing. Some solvers do have more overhead than others though, the extremes being CGS vs GMRES.

upsj commented on benchmark/utils/loggers.hpp:
```diff
@@ -229,4 +229,29 @@ struct ResidualLogger : gko::log::Logger {
+                               const gko::LinOp *, const gko::LinOp *,
+                               const gko::LinOp *) const override
+    {
+        this->num_iters = num_iterations;
```
When we have multiple right-hand sides: Do I understand correctly that this logs the maximum number of iterations over all rhs?

yhmtsai commented on benchmark/utils/overhead_linop.hpp:
```diff
@@ -0,0 +1,196 @@
+    friend class EnablePolymorphicObject<Overhead, LinOp>;
+
+public:
+    using value_type = ValueType;
```
```suggestion
```
`value_type` is unused here.

tcojean replied:
I removed it here since that is not something exposed to users, but as a side note these are usually unused in the code's class itself. They can be used from outside the class.

yhmtsai replied:
You are right. I forgot that it can be used from outside.
I am okay for keeping or removing it.

tcojean commented on benchmark/utils/overhead_linop.hpp:
```diff
@@ -0,0 +1,196 @@
+#ifndef GKO_BENCHMARK_UTILS_OVERHEAD_LINOP_
```
Thanks, the `make format_header` worked as expected.
